pkg/controller/healthzcheck/controller.go

```diff
@@ -0,0 +1,188 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package healthzcheck
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/record"
+	unversioned_core "k8s.io/kubernetes/pkg/client/typed/generated/core/unversioned"
+	"k8s.io/kubernetes/pkg/probe"
+	http_probe "k8s.io/kubernetes/pkg/probe/http"
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+// APIServerHealthChecker checks the health of API servers,
+// using the server address and ports provided.
+type APIServerHealthChecker interface {
+	Check(masterAddress string, masterPorts []api.EndpointPort) (bool, error)
+}
+
+// httpsHealthzChecker checks API server health by issuing a GET request over HTTPS to
+// the API server at the configured "healthz" path (e.g. "/healthz").  In order for this
+// to work, the caller (the controller manager) must be able to reach all of the API servers.
+// It searches for HTTP ports if no HTTPS ports are available.
+type httpsHealthzChecker struct {
+	healthzPath string
+	client      *http.Client
+}
+
+func (c *httpsHealthzChecker) choosePort(ports []api.EndpointPort) (int, string, error) {
+	healthPort := -1
+	healthProto := ""
+	for _, port := range ports {
+		// use the HTTP port if necessary, but let any HTTPS ports override it
+		if healthPort == -1 && port.Name == "http" && port.Protocol == api.ProtocolTCP {
+			healthPort = port.Port
+			healthProto = "http"
+			// continue on to check if any HTTPS ports exist further down
+		} else if port.Name == "https" && port.Protocol == api.ProtocolTCP {
+			healthPort = port.Port
+			healthProto = "https"
+			break
+		}
+	}
+
+	if healthPort == -1 {
+		return -1, "", fmt.Errorf("no TCP port was called 'https' or 'http'")
+	}
+
+	return healthPort, healthProto, nil
+}
+
+// Check checks the health of the given API server by issuing a GET request over HTTPS
+// to the given address (using the 'https' port, or 'http' port if necessary)
+func (c *httpsHealthzChecker) Check(masterAddress string, masterPorts []api.EndpointPort) (bool, error) {
+	if len(masterPorts) == 0 {
+		return false, fmt.Errorf("unable to determine port to use to check API server %s healthz: no ports provided", masterAddress)
+	}
+
+	masterPort, masterProto, err := c.choosePort(masterPorts)
+	if err != nil {
+		return false, fmt.Errorf("unable to determine port to use to check API server %s healthz URL: %v", masterAddress, err)
+	}
+	urlParsed, err := url.Parse(fmt.Sprintf("%s://%s:%v/%s", masterProto, masterAddress, masterPort, c.healthzPath))
+	if err != nil {
+		return false, err
+	}
+	maxRetries := 4
+	retries := 0
+	delay := 125 * time.Millisecond
+	for retries < maxRetries {
```

Comment:
This may have predated ExponentialBackoff

On Mon, Apr 25, 2016 at 3:49 PM, Solly Ross <notifications@github.com>
wrote:

> In pkg/controller/healthzcheck/controller.go
> <https://github.com/kubernetes/kubernetes/pull/20975#discussion_r60976249>
> :
>
> > +	if len(masterPorts) == 0 {
> > +		return false, fmt.Errorf("unable to determine port to use to check API server %s healthz: no ports provided", masterAddress)
> > +	}
> > +
> > +	masterPort, masterProto, err := c.choosePort(masterPorts)
> > +	if err != nil {
> > +		return false, fmt.Errorf("unable to determine port to use to check API server %s healthz URL: %v", masterAddress, err)
> > +	}
> > +	urlParsed, err := url.Parse(fmt.Sprintf("%s://%s:%v/%s", masterProto, masterAddress, masterPort, c.healthzPath))
> > +	if err != nil {
> > +		return false, err
> > +	}
> > +	maxRetries := 4
> > +	retries := 0
> > +	delay := 125 * time.Millisecond
> > +	for retries < maxRetries {
>
> I don't think you asked that before. I'll take a look (I have to page this
> back into RAM :-P).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20975/files/05374c7b831ab3f2ceaaff6f8234dce8c408dde4#r60976249>
>